Cubical/HITs/CumulativeHierarchy/Base.agda

```diff
@@ -0,0 +1,672 @@
+------------
+
+-- set up the basic hierarchy definition and _∈_ as recursive, higher inductive types
+data V (ℓ : Level) : Type (ℓ-suc ℓ)
+_∈_ : (S T : V ℓ) → hProp (ℓ-suc ℓ)
+
+eqImage : {X Y : Type ℓ} (ix : X → V ℓ) (iy : Y → V ℓ) → Type (ℓ-suc ℓ)
+eqImage {X = X} {Y = Y} ix iy = (∀ (a : X) → ∥ Σ[ b ∈ Y ] (iy b) ≡ (ix a) ∥) ⊓′ (∀ (b : Y) → ∥ Σ[ a ∈ X ] (ix a) ≡ (iy b) ∥)
+
+data V ℓ where
+  sett : (X : Type ℓ) → (X → V ℓ) → V ℓ
+  seteq : (X Y : Type ℓ) (ix : X → V ℓ) (iy : Y → V ℓ) (eq : eqImage ix iy) → sett X ix ≡ sett Y iy
+  setIsSet : isSet (V ℓ)
+
+A ∈ sett X ix = ∥ Σ[ i ∈ X ] (ix i ≡ A) ∥ₚ
+A ∈ seteq X Y ix iy (f , g) i =
+  ⇔toPath {P = A ∈ sett X ix} {Q = A ∈ sett Y iy}
+    (λ ax → do (x , xa) ← ax ; (y , ya) ← f x ; ∣ y , ya ∙ xa ∣)
+    (λ ay → do (y , ya) ← ay ; (x , xa) ← g y ; ∣ x , xa ∙ ya ∣) i
+  where open PropMonad
+A ∈ setIsSet a b p q i j = isSetHProp (A ∈ a) (A ∈ b) (λ j → A ∈ p j) (λ j → A ∈ q j) i j
+
+-- a provisional definition of equality, we will show that this is actually equivalent to an hProp in universe ℓ
+infix 4 _≡ₛ_
+_≡ₛ_ : (a b : V ℓ) → hProp (ℓ-suc ℓ)
+_≡ₛ_ a b = (a ≡ b) , setIsSet a b
+
+-- setup a general eliminator into h-sets
+record ElimSet {Z : (s : V ℓ) → Type ℓ'} (isSetZ : ∀ s → isSet (Z s)) : Type (ℓ-max ℓ' (ℓ-suc ℓ)) where
+  field
+    ElimSett : (X : Type ℓ) (ix : X → V ℓ)
+             -- ^ the structural parts of the set
+             → (rec : ∀ x → Z (ix x))
+             -- ^ a recursor into the elements
+             → Z (sett X ix)
+    ElimEq   : (X₁ X₂ : Type ℓ) (ix₁ : X₁ → V ℓ) (ix₂ : X₂ → V ℓ) (eq : eqImage ix₁ ix₂)
+             -- ^ the structural parts of the seteq path
+             → (rec₁ : ∀ x₁ → Z (ix₁ x₁)) (rec₂ : ∀ x₂ → Z (ix₂ x₂))
+             -- ^ recursors into the elements
+             → ((x₁ : X₁) → ∥ Σ[ x₂ ∈ X₂ ] (Σ[ p ∈ (ix₂ x₂ ≡ ix₁ x₁) ] PathP (λ i → Z (p i)) (rec₂ x₂) (rec₁ x₁)) ∥)
+             → ((x₂ : X₂) → ∥ Σ[ x₁ ∈ X₁ ] (Σ[ p ∈ (ix₁ x₁ ≡ ix₂ x₂) ] PathP (λ i → Z (p i)) (rec₁ x₁) (rec₂ x₂)) ∥)
+             -- ^ proofs that the recursors have equal images
+             → PathP (λ i → Z (seteq X₁ X₂ ix₁ ix₂ eq i)) (ElimSett X₁ ix₁ rec₁) (ElimSett X₂ ix₂ rec₂)
+
+module _ {Z : (s : V ℓ) → Type ℓ'} {isSetZ : ∀ s → isSet (Z s)} (E : ElimSet isSetZ) where
+  open ElimSet E
+  elim : (s : V ℓ) → Z s
+  {-# TERMINATING #-} -- TODO: remove terminating ???
+  elim (sett X ix) = ElimSett X ix (elim ∘ ix)
+  elim (seteq Y₁ Y₂ iy₁ iy₂ eq i) = ElimEq Y₁ Y₂ iy₁ iy₂ eq (elim ∘ iy₁) (elim ∘ iy₂) rec₁→₂ rec₂→₁ i where
+    open PropMonad
+    rec₁→₂ : (y₁ : Y₁) → ∥ Σ[ y₂ ∈ Y₂ ] (Σ[ p ∈ (iy₂ y₂ ≡ iy₁ y₁) ] PathP (λ i → Z (p i)) (elim (iy₂ y₂)) (elim (iy₁ y₁))) ∥
+    rec₂→₁ : (y₂ : Y₂) → ∥ Σ[ y₁ ∈ Y₁ ] (Σ[ p ∈ (iy₁ y₁ ≡ iy₂ y₂) ] PathP (λ i → Z (p i)) (elim (iy₁ y₁)) (elim (iy₂ y₂))) ∥
+                                                        --v-------v-- the problematic part for the termination checker
+    rec₁→₂ y₁ = do (y₂ , yy) ← eq .fst y₁ ; ∣ y₂ , yy , cong elim yy ∣
```

Comment:
So the part that the termination checker doesn't like is `cong elim yy`? Did you try unfolding the definition of cong?

Reply:
I tried to do so, but the error remains (line numbers not accurate since I moved stuff at the top)

```
/home/martin/src/cubical/Cubical/HITs/CumulativeHierarchy/Base.agda:140,3-152,71
Termination checking failed for the following functions:
  _.elim
Problematic calls:
  λ { (y₂ , yy) → ∣ y₂ , yy , (λ i → elim (yy i)) ∣ }
  Cubical.HITs.CumulativeHierarchy.Base.rec₁→₂
    (at /home/martin/src/cubical/Cubical/HITs/CumulativeHierarchy/Base.agda:143,87-93)
  elim (yy i)
    (at /home/martin/src/cubical/Cubical/HITs/CumulativeHierarchy/Base.agda:148,64-68)
```

I read this as if the termination checker sees two general arguments `y₂ , yy` in the lamba passed to `Prop.rec` (after desuraging do syntax), and a call `elim (yy i)`. It does not realize that these are covered since all actually occuring values come from eliminating `eq .fst y₁` which is covered.

Reply:
yeah, the termination checker might be able to see through this recursion pattern if you define a local function that pattern matches on the result of `eq .fst y₁`.

Reply:
@Saizan local function definition worked, very cool.

I tried using `with`-abstraction first and hit a "type-checking for with-abstractions is not implemented" when trying something like

    rec₁→₂ x₁ with eq .fst x₁
    rec₁→₂ x₁ | ∣ x₂ , xx ∣ = ∣ x₂ , xx , (λ i → elim (xx i)) ∣
    rec₁→₂ x₁ | squash a b i = squash (rec₁→₂ x₁ | a) (rec₁→₂ x₁ | b) i

but using a local definition got rid of the pragma, i.e. this module is now `--safe` with my next push!